### PR TITLE
Add activity callback function

### PR DESCRIPTION
**Pull request by Ing-Dom**
Add the possibility to have a callback function signaling knx activity (for e.g. flashing LEDs)

Can be enabled with KNX_ACTIVITYCALLBACK

### CODE REVIEW

**Comment by thelsing** [comment]
For achitectional reasons I will not merge this. The knx-Facade implements the facade pattern an simplifies usage for lib users. It is completely optional. Therefore you can't use it in any of the other parts of the stack. For the same reason its global instance must not be used in the stack. 

The correct way is to add the callback to be Bau. There can be multiple Baus (in routers for example) and each can have its own callback.

**Review by thelsing — CHANGES_REQUESTED** [review]
See comment on PR:

**Comment by Ing-Dom** [comment]
yes you are right.
I'll rework it.
do you really mean bau, because the router baus in the stack are "single"
I think you mean more then one link layer do you, like in the bau091A ? (or the other coupler baus)

**Comment by Ing-Dom** [comment]
Hi Thomas,

I redesigned the feature according you hints.

**Comment by thelsing** [comment]
I mean more than one logical device in one physical device. My ip router for example has one PA for the router application and a different one for additional stuff like providing time to the bus. The facade was never meant for bigger stuff like openknx with people who know how the stack is structured,  but for simple sketches for typical arduino users.

File: src/knx/data_link_layer.h [hunk 1]
```diff
@@ -10,6 +10,18 @@
...
 
+typedef void (*ActivityCallback)(uint8_t info);
+
+class DataLinkLayerCallbacks
+{
+protected:
+    ActivityCallback _activityCallback = nullptr;
+public:
+    virtual ~DataLinkLayerCallbacks() = default;
+    virtual void Activity(uint8_t info);
```
Comment (thelsing) [inline src/knx/data_link_layer.h:21]: Methods should start lower case.

File: src/knx/ip_data_link_layer.cpp [hunk 2]
```diff
@@ -67,6 +76,10 @@ void IpDataLinkLayer::loop()
             KnxIpSearchResponse searchResponse(_ipParameters, _deviceObject);
 
             auto hpai = searchRequest.hpai();
+#ifdef KNX_ACTIVITYCALLBACK
+    if(_dllcb)
```
Comment (thelsing) [inline src/knx/ip_data_link_layer.cpp:80]: Indent looks of. Is it a spaces vs. tabs issue?

File: src/knx/tpuart_data_link_layer.cpp [hunk 3]
```diff
@@ -653,6 +658,10 @@ bool TpUartDataLinkLayer::sendSingleFrameByte()
     if (_TxByteCnt >= _sendBufferLength)
     {
         _TxByteCnt = 0;
+#ifdef KNX_ACTIVITYCALLBACK
+    if(_dllcb)
```
Comment (thelsing) [inline src/knx/tpuart_data_link_layer.cpp:662]: Same indent format as above.

File: src/knx/tpuart_data_link_layer.h [hunk 4]
```diff
@@ -21,8 +21,12 @@ class TpUartDataLinkLayer : public DataLinkLayer
     using DataLinkLayer::_platform;
 
   public:
+    // TpUartDataLinkLayer(DeviceObject& devObj, NetworkLayerEntity& netLayerEntity,
```
Comment (thelsing) [inline src/knx/tpuart_data_link_layer.h:24]: Is there any reason to keep this?

**Review by thelsing — CHANGES_REQUESTED** [review]
This version is fine beside some minor style issues.  Sorry for being nitpicky but I have to work with code at work were people weren't for a few years...

File: src/knx/tpuart_data_link_layer.cpp (hunk 3 above)
Comment (Ing-Dom) [reply src/knx/tpuart_data_link_layer.cpp:662]: fixed

File: src/knx/ip_data_link_layer.cpp (hunk 2 above)
Comment (Ing-Dom) [reply src/knx/ip_data_link_layer.cpp:80]: fixed

File: src/knx/data_link_layer.h (hunk 1 above)
Comment (Ing-Dom) [reply src/knx/data_link_layer.h:21]: changed Activity to activity

**Comment by Ing-Dom** [comment]
> I mean more than one logical device in one physical device. My ip router for example has one PA for the router application and a different one for additional stuff like providing time to the bus. The facade was never meant for bigger stuff like openknx with people who know how the stack is structured, but for simple sketches for typical arduino users.

I work on an ip router (091A) based on this stack for alomost one year (PR will come if its a little more field tested).
For this I thought about how to solve the second application (virtual knx device) but I think that will not be so easy - some sort of bus-access splitter (or however you call it) would be neccessary, right?

**Review by thelsing — APPROVED** [review]
All ok now.

**Comment by thelsing** [comment]
> I work on an ip router (091A) based on this stack for alomost one year (PR will come if its a little more field tested). For this I thought about how to solve the second application (virtual knx device) but I think that will not be so easy - some sort of bus-access splitter (or however you call it) would be neccessary, right?

For knx-ip you should be able open more than one multicast connection from one host and use them in parallel.

**Pull request merged**